proposals/0351-regex-builder.md

```diff
@@ -428,16 +442,226 @@ extension RegexComponentBuilder {
+
+```swift
+extension Capture {
+  public init<R: RegexComponent, W>(
+    _ component: R
+  ) where Output == (Substring, W), R.RegexOutput == W
+  
+  public init<R: RegexComponent, W>(
+    _ component: R, as reference: Reference<W>
+  ) where Output == (Substring, W), R.RegexOutput == W
+  
+  public init<R: RegexComponent, W, NewCapture>(
+    _ component: R,
+    transform: @escaping (Substring) throws -> NewCapture
+  ) where Output == (Substring, NewCapture), R.RegexOutput == W
+  
+  public init<R: RegexComponent, W>(
+    @RegexComponentBuilder _ component: () -> R
+  ) where Output == (Substring, W), R.RegexOutput == W
+
+  // ... `O(arity)` overloads
+}
+  
+extension TryCapture {
+  public init<R: RegexComponent, W, NewCapture>(
+    _ component: R,
+    transform: @escaping (Substring) throws -> NewCapture?
+  ) where Output == (Substring, NewCapture), R.RegexOutput == W
+  
+  public init<R: RegexComponent, W, NewCapture>(
+    @RegexComponentBuilder _ component: () -> R,
+    transform: @escaping (Substring) throws -> NewCapture?
+  ) where Output == (Substring, NewCapture), R.RegexOutput == W
+
+  // ... `O(arity)` overloads
+}
+```
+
+</details>
+
+### Reference
+
+Reference is a feature that can be used to achieve named captures and named backreferences from textual regexes. Simply state what type the reference will hold on to and you can use it later once you've matched a string to get back a specific capture. Note the type you pass to reference will be whatever the result of a capture's transform is. A capture with no transform always has a reference type of `Substring`.
+
+```swift
+let kind = Reference(Substring.self)
+
+let regex = Capture(as: kind) {
+  ChoiceOf {
+    "CREDIT"
+    "DEBIT"
+  }
+}
+
+let input = "CREDIT"
+if let result = regex.firstMatch(in: input) {
+  print(result[kind]) // Optional("CREDIT")
+}
+```
+
+Capturing stores the most recently captured content, and references can be used as a name to look up the result of matching. The reference itself can also be used within a regex (commonly called a "backreference") to match the most recently captured content during matching.
+
+```swift
+let a = Reference(Substring.self)
+let b = Reference(Substring.self)
+let c = Reference(Substring.self)
+let regex = Regex {
+  Capture("abc", as: a)
+  Capture("def", as: b)
+  ZeroOrMore {
+    Capture("hij", as: c)
+  }
+  a
+  Capture(b)
+}
+
+if let result = regex.firstMatch(in: "abcdefabcdef") {
+  print(result[a]) // => Optional("abc")
+  print(result[b]) // => Optional("def")
+  print(result[c]) // => nil
+}
+```
+
+A regex is considered invalid when it contains a use of reference without it ever being captured as (used as `Capture(..., as: ...)` or `Capture(as: ...) { ... }`) in the regex. When this occurs in the regex builder DSL, a runtime error will be reported. Similarly, the argument to a `Regex.Match.subscript(_:)` must have been captured as in the regex that produced the match.
```

Comment:
I still think this language conflates two meanings of "captured". The meaning here is that it needs to be passed in a call to Capture or TryCapture, but the other way we use "captured" is that it matches some portion (even an empty range) of the input text. Can we make that more clear, here and in the subscript docs?